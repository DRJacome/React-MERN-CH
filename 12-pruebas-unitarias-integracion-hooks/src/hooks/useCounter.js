import { useState } from "react"

export const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState)

  const reset = () => {
    setCounter(initialState)
  }

  const incrementar = () => {
    setCounter(counter + 1)
  }
  const decrementar = () => {
    setCounter(counter - 1)
  }

  return {
    state: counter,
    incrementar,
    decrementar,
    reset
  }
}