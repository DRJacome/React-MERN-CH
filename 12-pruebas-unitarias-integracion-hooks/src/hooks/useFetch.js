import { useEffect, useRef, useState } from "react"

export const useFetch = (url) => {
  const isMounted = useRef(true)
  const [estado, setEstado] = useState({ data: null, loading: true, error: null });

  useEffect(() => {
    return () => {
      isMounted.current = false
    }
  }, [])

  useEffect(() => {
    setEstado({ data: null, loading: true, error: null })
    fetch(url)
      .then(respuesta => respuesta.json())
      .then(data => {
        if (isMounted) {
          setTimeout(() => {
            setEstado({
              loading: false,
              error: null,
              data
            });
          }, 100);
        } else {
          console.log('setState() no se llamó');
        }
      })
      .catch(error => {
        setEstado({
          data: null,
          loading: false,
          error: "No se pudo obtener la información"
        })
      });
  }, [url])

  return estado;
}
