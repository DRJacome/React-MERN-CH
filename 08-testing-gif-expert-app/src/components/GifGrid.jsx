import React from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from "./GifGridItem"
//import { getGifs } from "../helpers/getGifs"

export const GifGrid = ({ categoria }) => {
  const { data: imagenes, loading } = useFetchGifs(categoria)

  /*   useEffect(() => {
      getGifs(categoria)
        .then(
        )
      /*  El segundo argumento del useEffect es llamado "Lista/array de dependencias []" y 
      * y objetivo es controlar cuándo se renderiza el componente. Si el array está vacío,
      * el componente sólo se renderizará por primera vez, evitando que se renderice continuamente.
      
    }, [categoria])
   */

  return (
    <>
      <h3 className='animate__animated animate__fadeIn'>{categoria}</h3>
      {loading && <p className='carta animate__animated animate__flash'>Cargando</p>}
      {<div className="contenedorCartas">
        {imagenes.map(img => (
          <GifGridItem
            key={img.id}
            {...img}
          />
        ))}
      </div>}
    </>
  )
}

GifGrid.propTypes = {
  categoria: PropTypes.string.isRequired
}