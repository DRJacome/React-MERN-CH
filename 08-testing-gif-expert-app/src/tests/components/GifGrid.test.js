import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import GifGrid from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')


describe('Pruebas sobre <GifGrid/>', () => {
  const categoria = 'One Punch Man'

  test('Debe mostrar el componente correctamente (snapshot)', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    })

    const wrapper = shallow(<GifGrid categoria={categoria} />);
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe mostrar items cuando se cargan imÃ¡genes de useFetchGifs', () => {
    const gifs = [{
      id: 'ABC',
      url: 'https://localhost/algo/algo.jpg',
      title: 'Cualquier cosa'
    },
    {
      id: '123',
      url: 'https://localhost/alguno/alguno.jpg',
      title: 'Cualquier item'
    }];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    })
    const wrapper = shallow(<GifGrid categoria={categoria} />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('p').exists()).toBe(false)
    expect(wrapper.find('GifGridItem').length).toBe(gifs.length)
  })
})