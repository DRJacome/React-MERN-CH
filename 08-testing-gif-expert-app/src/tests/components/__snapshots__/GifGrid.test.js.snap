// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <GifGrid/> Debe mostrar el componente correctamente (snapshot) 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch Man
  </h3>
  <p
    className="carta animate__animated animate__flash"
  >
    Cargando
  </p>
  <div
    className="contenedorCartas"
  />
</Fragment>
`;

exports[`Pruebas sobre <GifGrid/> Debe mostrar items cuando se cargan im√°genes de useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch Man
  </h3>
  <div
    className="contenedorCartas"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/algo/algo.jpg"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Cualquier item"
      url="https://localhost/alguno/alguno.jpg"
    />
  </div>
</Fragment>
`;
